pkg/nodeagent/controller/healthcheck/add.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthcheck
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/namespaces"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/gardener/gardener/pkg/nodeagent/dbus"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "healthcheck"
+
+const defaultIntervalSeconds = 30
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName)
+	}
+	if r.DBus == nil {
+		r.DBus = dbus.New()
+	}
+	if len(r.HealthCheckers) == 0 {
+		err := r.setDefaultHealthChecks()
+		if err != nil {
+			return err
+		}
+	}
+	if r.HealthCheckIntervalSeconds == 0 {
+		r.HealthCheckIntervalSeconds = defaultIntervalSeconds
+	}
+
+	node := &metav1.PartialObjectMetadata{}
+	node.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Node"))
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(node).
```

Comment:
```suggestion
		For(node, builder.WithPredicates(nodePredicate)).
```

ref https://github.com/gardener/gardener/blob/c8c174de728a28c6ed83a292dfc5a2726138e7c2/pkg/nodeagent/controller/node/add.go#L50